### ISSUE TITLE
Validator key assignment ('key delegation')

### ISSUE THREAD

**Issue by AdityaSripal** [cosmos/interchain-security#26]
Allow Validators to set a different consensus key for each baby chain

- [ ] Create Msg type
- [ ] Implement setting handler (should always override previous value)
- [ ] Default to same consensus key as parent chain
- [ ] In Endblocker, make the transformation from parent keys to baby keys in the change set before sending the packet

**Comment by mpoke** [comment]
This will be a user tx on the provider CCV module, i.e., CreateKey(valAddr, consumerChainId).

**Comment by jtremback** [comment]
Jack's comment- a validator set change of more than 1/3 will break all the IBC light client

**Comment by mpoke** [comment]
I'm not sure I understand. I thought that the validators can choose their consensus keys for a given chain before the chain is initialized. Can validators do this while validating on the consumer chain? What would be the benefit of that?

**Comment by jackzampolin** [comment]
Key rotation is a nice to have feature but could present problems. Setting the key prior to consumer chain start should be fine.

**Comment by mpoke** [comment]
@okwme Is this feature something we need for V1 or it's a "nice to have"?

**Comment by mohammedpatla** [comment]
> Key rotation is a nice to have feature but could present problems. Setting the key prior to consumer chain start should be fine.

Honestly waiting for key rotation option.
If this was available it would be more open to reusing priv key across child chains. For now we spin a new one for every chain.

**Comment by marbar3778** [comment]
We are thinking of providing the expected_keeper interface for the ccv module as such:

```
type LookUpModule interface {
 GetLookupList() []entry
 GetUserEntry(address sdk.Address) entry 
}
```

Does this make sense from the ccv point of view?


(names aren't final)

**Comment by marbar3778** [comment]
QQ: If a validator joins the set how long would it take the update to propagate to N chains? The issue arises that once a validator joins the active set how long do they have to start spinning up on consumer chains, this would be the deciding factor for how this gets implemented for me

**Comment by okwme** [comment]
If a validator joins the set, this should trigger an update to the child chain so it could be quite fast. Faster than the val might have to submit a new key for the child chain. However, maybe it shoudl be possible to associate keys before joining the val set? Also probably shouldn't try to validate whether a key is or is not currently a validator for the same reason. Would this improve the scenario you were worried about @marbar3778 ?

**Comment by marbar3778** [comment]
yea that would solve it. I guess the docs would say when becoming a validator if you have not registered a child chain consensus key we will use the default one. The outstanding question is if a validator changes it after the validator is created on the child chain are there security assumptions to think about?

**Comment by marbar3778** [comment]
easiest would be fail on create-validator message if there are no entries for that validator in the key lookup table

**Comment by okwme** [comment]
I think this is solved if there was a store with Keys of operator PKs and Values of operator PKs. Provider Chain validators would register (or not) their consumer chain PK. These are used to send voting power to the consumer over CCV. A provider operator can change the consumer PK as often as they want. Once updated, they will need to use this new key as part of their node as soon as the update packet has landed on the child chain (they could just as well run two nodes with each PK as one will be ignored until the change happens without risk of double signing). If they consumer chain sends rewards to the parent chain but the parent has recently updated their PK, this is also not a problem, as the previous consumer PK is still a Key in the store pointing to the provider PK. It will be redundant with the new consumer PK that is also pointing to the provider PK, but that's not a problem. We may want to consider a garbage collection that is some period after it is assumed the new key has landed on the consumer chain.

the key of the store is for operators of both consumer and provider. The value is both consumer and provider. There is no collision on the store key even though there may be redundant entries in the store value.

**Comment by okwme** [comment]
> easiest would be fail on create-validator message if there are no entries for that validator in the key lookup table. This would mean we require validators to not use the same key.

how would this work for validators that are already in the set? Shouldn't we fail over to the same key if they did not register one?

The child chain won't begin producing blocks until 2/3 of the validators come online. It is same to assume that after a proposal passes, these validators won't turn on their nodes until after registering their PKs.

**Comment by okwme** [comment]
actually, there is no need to store consumer chain PKs in the store at all.
the only reason would be to associate a consumer PK with a provider PK in order to pay rewards, but this isn't necessary as the rewards go straight to the distribution module where the correct PKs are already present.

**Comment by mpoke** [comment]
> actually, there is no need to store consumer chain PKs in the store at all.
> the only reason would be to associate a consumer PK with a provider PK in order to pay rewards, but this isn't necessary as the rewards go straight to the distribution module where the correct PKs are already present.

@okwme We still need the mapping from consumer PKs to provider PKs to match `SlashPackets` to the validators on the provider.

**Comment by mpoke** [comment]
>  Provider Chain validators would register (or not) their consumer chain PK. These are used to send voting power to the consumer over CCV. A provider operator can change the consumer PK as often as they want.

This may lead to the problem pointed by Jack, i.e., https://github.com/cosmos/interchain-security/issues/26#issuecomment-1113623525. How does it work currently on a single chain? Can a validator change its PK while part of the validator set? 

I think for the first release, we should keep it simple and allow the validators to decide on their consumer chain PK only once before they join the validator set. By default is the same PK as on the provider.

**Comment by okwme** [comment]
if it's required then it's also fine to have a general look up table like i first described and the validator can change as often as they want to with no problem.

If the lookup table is initially like this:

| Key           | Value         |
|---------------|---------------|
| PK Provider A | PK Consumer A |
| PK Consumer A | PK Provider A |

and `PK Provider A` changes their Consumer PK it would just look like this afterwards:


| Key           | Value         |
|---------------|---------------|
| PK Provider A | PK Consumer B |
| PK Consumer A | PK Provider A |
| PK Consumer B | PK Provider A |


query is always up to date

**Comment by mpoke** [comment]
Yeah, but on the consumer chain, we may have the case that in two subsequent blocks the validators corresponding to more than 1/3 of the voting power change their PKs. As a result, the client to the consumer chain can no longer be updated. 

My question still stands:
> In the current version of SDK, can a validator change its PK while part of the validator set?

**Comment by mpoke** [comment]
After discussions within the team, the consensus was to allow every validator to change its consensus key on any consumer chain at any point in time. The issue discussed in https://github.com/cosmos/interchain-security/issues/26#issuecomment-1113623525 can be address by the relayer switching to the sequential protocol for updating the client.

**Comment by danwt** [comment]
Presumably the mapping should not be opaque? It seems likely that people will want to know which validators on the consumers map to which provider validators and vice versa. Thus we'll need queries/RPC for this mapping, and perhaps additional metadata anyways on the consumer for each validator.

Have I got it right?

**Comment by danwt** [comment]
I made a prototype design, it's a WIP.

- https://github.com/cosmos/interchain-security/issues/360

It's entirely provider side, the consumer doesn't need to know anything.

**Comment by danwt** [comment]
TODOs

- https://github.com/cosmos/interchain-security/issues/503
- test the scenarios in #503 by finishing the tests which are commented out
  - https://github.com/cosmos/interchain-security/blob/7616700b6ec52e264ea2966cf364395d95673ee8/x/ccv/provider/keeper/key_assignment_test.go#L803-L807
  - https://github.com/cosmos/interchain-security/blob/7616700b6ec52e264ea2966cf364395d95673ee8/x/ccv/provider/keeper/key_assignment_test.go#L646-L653
  - https://github.com/cosmos/interchain-security/blob/7616700b6ec52e264ea2966cf364395d95673ee8/x/ccv/provider/keeper/key_assignment_test.go#L709-L716
  - bring back docs ect from this branch https://github.com/cosmos/interchain-security/tree/danwt/key-assignment-v4/docs/key-assignment

@okwme here is a quick summary. I will flesh out tomorrow morning. I recommend reading [this doc](https://github.com/cosmos/interchain-security/blob/danwt/key-assignment-v4/docs/key-assignment/design.md) to figure out what is going on.
